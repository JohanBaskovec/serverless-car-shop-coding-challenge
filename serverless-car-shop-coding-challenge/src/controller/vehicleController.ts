import { type CreateVehicleDTO, type UpdateVehicleDTO } from '../model/dto/vehicleDTO.js';
import { CrudController } from './crudController.js';
import { type VehicleDocument } from '../generated/mongoose.gen.js';

export class VehicleController extends CrudController<VehicleDocument, CreateVehicleDTO, UpdateVehicleDTO> {
}
