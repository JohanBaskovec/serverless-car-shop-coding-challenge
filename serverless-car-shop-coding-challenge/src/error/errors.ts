export class MissingRequestBody extends Error {
    constructor() {
        super('Missing body in HTTP request.');
        this.name = 'MissingRequestBody';
    }
}

export class MissingKeyInJSONBodyError extends Error {
    constructor(private readonly _missingKeys: string[]) {
        super(`Missing key in JSON: ${_missingKeys.join(',')}`);
        this.name = 'MissingKeyInJSONBodyError';
    }

    get missingKeys(): string[] {
        return this._missingKeys;
    }
}

export class MissingPathParametersError extends Error {
    constructor(private readonly _missingParameters: string[]) {
        super(`Missing path parameters: ${_missingParameters.join(',')}`);
        this.name = 'MissingPathParametersError';
    }

    get missingParameters(): string[] {
        return this._missingParameters;
    }
}

export class EntityDoesNotExistError extends Error {
    constructor(message: string) {
        super(message);
        this.name = 'EntityDoesNotExistError';
    }
}

export class ConstraintError extends Error {
    constructor(message: string) {
        super(message);
        this.name = 'ConstraintError';
    }
}
