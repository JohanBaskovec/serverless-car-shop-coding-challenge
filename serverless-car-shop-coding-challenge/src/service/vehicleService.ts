import { Vehicle } from '../model/vehicle.js';
import { type CreateVehicleDTO, type UpdateVehicleDTO } from '../model/dto/vehicleDTO.js';
import { EntityDoesNotExistError, MissingKeyInJSONBodyError } from '../error/errors.js';
import { type DealerService } from './dealerService.js';
import { type DealerDocument, type VehicleDocument } from '../generated/mongoose.gen.js';
import mongoose from 'mongoose';
import { type CrudService } from './crudService.js';
import { type APIGatewayProxyEventQueryStringParameters } from 'aws-lambda/trigger/api-gateway-proxy.js';
import { type DeleteResponse } from '../model/response.js';

export class VehicleService implements CrudService<VehicleDocument, CreateVehicleDTO, UpdateVehicleDTO> {
    private _dealerService: DealerService;

    get dealerService(): DealerService {
        return this._dealerService;
    }

    set dealerService(value: DealerService) {
        this._dealerService = value;
    }

    async getAll(queryStringParameters: APIGatewayProxyEventQueryStringParameters | null = null): Promise<VehicleDocument[]> {
        const params: any = {};
        if (queryStringParameters != null) {
            if (queryStringParameters.dealer != null) {
                params.dealer = queryStringParameters.dealer;
            }
        }
        return await Vehicle.find(params).exec();
    }

    async create(dto: CreateVehicleDTO): Promise<VehicleDocument> {
        const vehicle = await new Vehicle(dto).save();
        await this.dealerService.addVehicle(vehicle.dealer, vehicle, false);
        return vehicle;
    }

    async update(dto: UpdateVehicleDTO): Promise<VehicleDocument> {
        if (dto._id == null) {
            throw new MissingKeyInJSONBodyError(['_id']);
        }
        const vehicle = await Vehicle.findOne({ _id: dto._id }).exec();
        if (vehicle == null) {
            throw new EntityDoesNotExistError(`Vehicle with id ${dto._id} does not exist.`);
        }
        if (dto.dealer != null) {
            await this.setDealer(vehicle, new mongoose.Types.ObjectId(dto.dealer), true);
        }
        delete dto.dealer;
        const newVehicle = await Vehicle.findOneAndUpdate({ _id: dto._id }, dto, { new: true, runValidators: true }).exec();
        if (newVehicle == null) {
            throw new EntityDoesNotExistError(`Vehicle with id ${dto._id} does not exist.`);
        }
        return newVehicle;
    }

    async remove(id: string): Promise<DeleteResponse> {
        const vehicle = await Vehicle.findOne({ _id: id }).exec();
        if (vehicle == null) {
            throw new EntityDoesNotExistError(`Vehicle with id ${id} does not exist.`);
        }
        await this.dealerService.removeVehicle(vehicle.dealer, vehicle);
        return await Vehicle.deleteOne({ _id: id }).exec();
    }

    private async getVehicleDocument(vehicleOrVehicleId: VehicleDocument | mongoose.Types.ObjectId | string): Promise<VehicleDocument> {
        if (vehicleOrVehicleId instanceof mongoose.Types.ObjectId || typeof vehicleOrVehicleId === 'string') {
            const newVehicle = await Vehicle.findOne({ _id: vehicleOrVehicleId }).exec();
            if (newVehicle == null) {
                throw new EntityDoesNotExistError(`Vehicle with id ${vehicleOrVehicleId.toString()} does not exist.`);
            }
            return newVehicle;
        } else {
            return vehicleOrVehicleId;
        }
    }

    async setDealer(
        vehicleOrId: VehicleDocument | mongoose.Types.ObjectId,
        dealerOrId: DealerDocument | mongoose.Types.ObjectId,
        updateOppositeSide = false
    ): Promise<void> {
        const vehicle = await this.getVehicleDocument(vehicleOrId);

        await this.dealerService.removeVehicle(vehicle.dealer, vehicle);
        if (updateOppositeSide) {
            await this.dealerService.addVehicle(dealerOrId, vehicle, false);
        }

        vehicle.dealer = dealerOrId;
        await vehicle.save();
    }
}
