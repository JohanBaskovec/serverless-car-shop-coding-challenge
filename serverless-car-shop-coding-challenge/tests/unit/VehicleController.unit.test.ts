import mongoose from 'mongoose';
import { EntityDoesNotExistError, MissingPathParametersError } from '../../src/error/errors.js';
import { type VehicleService } from '../../src/service/vehicleService.js';
import { VehicleController } from '../../src/controller/vehicleController.js';

describe('vehicleController', () => {
    describe('getAll', () => {
        test('should return vehicles', async() => {
            const service = {
                getAll: jest.fn().mockReturnValue([])
            } as any as VehicleService;
            const controller = new VehicleController(service);
            const response = await controller.getAll();
            expect(service.getAll).toHaveBeenCalledTimes(1);
            expect(response).toMatchSnapshot();
        });
    });
    describe('create', () => {
        test('should create vehicle when request is valid', async() => {
            const fakeNewVehicle = {
                id: 'vehicle-id',
                name: 'vehicle-name'
            };
            const service = {
                create: jest.fn().mockImplementation(() => {
                    return fakeNewVehicle;
                })
            } as any as VehicleService;
            const controller = new VehicleController(service);
            const response = await controller.create(JSON.stringify({}));
            expect(service.create).toHaveBeenCalledTimes(1);
            expect(response).toMatchSnapshot();
        });
        test('should return an error when request is invalid', async() => {
            const service = {
                create: jest.fn().mockImplementation(() => {
                    throw new mongoose.Error.ValidationError();
                })
            } as any as VehicleService;
            const controller = new VehicleController(service);
            const response = await controller.create(JSON.stringify({}));
            expect(service.create).toHaveBeenCalledTimes(1);
            expect(response).toMatchSnapshot();
        });
    });
    describe('update', () => {
        test('should update vehicle when request is valid', async() => {
            const fakeNewVehicle = {
                id: 'vehicle-id',
                name: 'vehicle-name'
            };
            const service = {
                update: jest.fn().mockImplementation(() => {
                    return fakeNewVehicle;
                })
            } as any as VehicleService;
            const controller = new VehicleController(service);
            const response = await controller.update(JSON.stringify({}));
            expect(service.update).toHaveBeenCalledTimes(1);
            expect(response).toMatchSnapshot();
        });
        test('should return an error when request is invalid', async() => {
            const service = {
                update: jest.fn().mockImplementation(() => {
                    throw new mongoose.Error.ValidationError();
                })
            } as any as VehicleService;
            const controller = new VehicleController(service);
            const response = await controller.update(JSON.stringify({}));
            expect(service.update).toHaveBeenCalledTimes(1);
            expect(response).toMatchSnapshot();
        });
    });
    describe('remove', () => {
        test('should return an error when id is missing', async() => {
            const service = {
                remove: jest.fn().mockImplementation(() => {
                    throw new MissingPathParametersError(['id']);
                })
            } as any as VehicleService;
            const controller = new VehicleController(service);
            const response = await controller.remove({});
            expect(service.remove).toHaveBeenCalledTimes(0);
            expect(response).toMatchSnapshot();
        });
        test('should remove vehicle when request is valid', async() => {
            const service = {
                remove: jest.fn().mockImplementation(() => {
                    return { name: 'response from Mango' };
                })
            } as any as VehicleService;
            const controller = new VehicleController(service);
            const response = await controller.remove({ id: 'an id' });
            expect(service.remove).toHaveBeenCalledTimes(1);
            expect(response).toMatchSnapshot();
        });
        test("should return an error when vehicle doesn't exist", async() => {
            const service = {
                remove: jest.fn().mockImplementation(() => {
                    throw new EntityDoesNotExistError("Vehicle doesn't exist");
                })
            } as any as VehicleService;
            const controller = new VehicleController(service);
            const response = await controller.remove({ id: 'an id' });
            expect(service.remove).toHaveBeenCalledTimes(1);
            expect(response).toMatchSnapshot();
        });
    });
});
