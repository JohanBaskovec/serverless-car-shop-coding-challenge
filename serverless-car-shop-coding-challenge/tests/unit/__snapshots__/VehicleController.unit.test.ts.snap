// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vehicleController create should create vehicle when request is valid 1`] = `
{
  "body": "{"data":{"id":"vehicle-id","name":"vehicle-name"}}",
  "statusCode": 200,
}
`;

exports[`vehicleController create should return an error when request is invalid 1`] = `
{
  "body": "{"error":{"type":"ValidationError","message":"Validation failed"}}",
  "statusCode": 400,
}
`;

exports[`vehicleController getAll should return vehicles 1`] = `
{
  "body": "{"data":[]}",
  "statusCode": 200,
}
`;

exports[`vehicleController remove should remove vehicle when request is valid 1`] = `
{
  "body": "{"data":{"name":"response from Mango"}}",
  "statusCode": 200,
}
`;

exports[`vehicleController remove should return an error when id is missing 1`] = `
{
  "body": "{"error":{"type":"MissingPathParametersError","message":"Missing path parameters: id"}}",
  "statusCode": 400,
}
`;

exports[`vehicleController remove should return an error when vehicle doesn't exist 1`] = `
{
  "body": "{"error":{"type":"EntityDoesNotExistError","message":"Vehicle doesn't exist"}}",
  "statusCode": 404,
}
`;

exports[`vehicleController update should return an error when request is invalid 1`] = `
{
  "body": "{"error":{"type":"ValidationError","message":"Validation failed"}}",
  "statusCode": 400,
}
`;

exports[`vehicleController update should update vehicle when request is valid 1`] = `
{
  "body": "{"data":{"id":"vehicle-id","name":"vehicle-name"}}",
  "statusCode": 200,
}
`;
